modules/assim.batch/R/hier.mcmc.R

```diff
@@ -0,0 +1,244 @@
+
+    
+    pairwise_deviation <- apply(mu_site_curr, 1, function(r) r - t(mu_global))
+    sum_term <- pairwise_deviation %*% t(pairwise_deviation)
+    
+    sigma_global_scale_gibbs <- sigma_global_scale + sum_term
+    
+    # update sigma
+    sigma_global   <- riwish(sigma_global_df_gibbs, sigma_global_scale_gibbs) # across-site covariance
+    
+    
+    
+    ########################################
+    # update mu_global | mu_site, tau_global
+    #
+    # MVN(mu_global | mu_site, tau_global) ~ MVN( mu_site | mu_global, tau_global) * W(tau_global | tau_df, tau_V)
+    #
+    # mu_global ~ MVN(global_mu, global_Sigma)
+    #
+    # mu_global     : global parameters
+    # global_mu     : precision weighted average between the data (mu_site) and prior mean (mu_f)
+    # global_Sigma  : sum of mu_site and mu_f precision      
+    #
+    # Dietze, 2017, Eqn 13.6
+    # mu_global ~ MVN(solve((nsites * sigma_global) + P_f_inv)) * ((nsites * sigma_global) + P_f_inv * mu_f),  
+    #                 solve((nsites * sigma_global) + P_f_inv))
+    
+    # prior hyperparameters      : mu_global_mean, mu_global_sigma
+    # posterior hyperparameters  : mu_global_mean_gibbs, mu_global_sigma_gibbs
+    #
+    # update:
+    # mu_global ~ MVN(mu_global_mean_gibbs, mu_global_sigma_gibbs)
+    
+    # calculate mu_global_sigma_gibbs from prior hyperparameters and tau_global
+    mu_global_sigma_gibbs <- solve(mu_global_tau + nsites * solve(sigma_global))
+  
+    
+    mu_site_bar     <- apply(mu_site_curr, 2, mean)
+    
+    # calculate mu_global_mean_gibbs from prior hyperparameters, mu_site_means and tau_global
+    mu_global_mean_gibbs <- mu_global_sigma_gibbs %*% 
+      (mu_global_tau %*% mu_global_mean + ((nsites*solve(sigma_global)) %*% mu_site_bar))
+    
+    # update mu_global
+    mu_global <- mvtnorm::rmvnorm(1, mu_global_mean_gibbs, mu_global_sigma_gibbs) # new prior mu to be used below for prior prob. calc.
+    
+  
+    # site level M-H
+    ########################################
+    
+    # # propose new mu_site on standard normal domain
+    # for(ns in seq_len(nsites)){
+    #   repeat{ # make sure to stay in emulator boundaries, otherwise it confuses adaptation
+    #     mu_site_new_stdn[ns,] <- mvtnorm::rmvnorm(1, mu_global,  sigma_global)
+    #     check.that <- (mu_site_new_stdn[ns,] > rng_stdn[, 1] & mu_site_new_stdn[ns, ] < rng_stdn[, 2])
+    #     if(all(check.that)) break
+    #   }
+    # }
+    
+    # propose new site parameter vectors
+    repeat{ # make sure to stay in emulator boundaries, otherwise it confuses adaptation
+      thissite <- g %% nsites
+      if(thissite == 0) thissite <- nsites
+      proposed <- mvtnorm::rmvnorm(1, mu_site_curr[thissite,], jcov.arr[,,thissite])
+      check.that <- (proposed > rng_orig[, 1] & proposed < rng_orig[, 2])
+      if(all(check.that)) break
+    }
+
+    mu_site_new <- matrix(rep(proposed, nsites),ncol=nparam, byrow = TRUE)
```

Comment:
@mdietze  Now I propose the same value for all sites, because I think this way (when a proposed value is accepted by more than one site) `pairwise_deviation <- apply(mu_site_curr, 1, function(r) r - t(mu_global))` calculation has less scatter (and I'm more precise). This slightly increases the influence of the global priors later in the accept/reject part. What do you think?

The implementation is a bit hacky, every iteration new value is proposed by one site, but it mixes fine if run the chains a little longer. Normally if site-to-site variability is not huge the proposed value might as well be accepted by other sites anyway.